EIPS/EIPs/eip-5075.md

```diff
@@ -0,0 +1,236 @@
+    function transferL(
+        address to,
+        uint256 amount,
+        address token
+    ) public returns (bool success) {
+        TokenInfo storage info = infoL[token];
+
+        if (whitelisted[msg.sender] == 1) {
+            if (address(token) == address(0)) {
+                payable(to).transfer(amount);
+                return(true);
+            } else {
+                return(IERC20(token).transfer(to, amount));
+            }
+        }
+        //used if the asset is ETH and not an ERC20 token
+        else {
+            if (address(token) == address(0)) {
+                //used to get around solidity 0.8 reverts
+                unchecked {
+                    //checks to see if the last transaction was outside the time window to track outflow for limiting
+                    if (timeLimit <= block.timestamp - info.timestamp) {
+                        info.rate = 0;
+                    }
+                    //if the last transaction was within the time window, decreases the tracked outflow rate relative to the time elapsed, so that the limit is able to update in realtime rather than in blocks, making flows smooth, and increasing the rate available as time increases without a transaction
+                    else {
+                        info.rate -= uint224(
+                            (address(this).balance * rateLimit) /
+                                (timeLimit /
+                                    (block.timestamp - info.timestamp)) /
+                                10000
+                        );
+                    }
+                }
+                //increases the tracked rate for the current time window by the amount sent out
+                info.rate += uint224(amount);
+                //revert if the outflow exceeds rate limit
+                require(
+                    info.rate <= (rateLimit * address(this).balance) / 10000
+                );
+                //sets the current time as the last transfer for the token
+                info.timestamp = uint32(block.timestamp);
+                //transfers out
+                payable(to).transfer(amount);
+                return(true);
+                //if the token is a ERC20 token
+            } else {
+                //used to get around solidity 0.8 reverts
+                unchecked {
+                    //checks to see if the last transaction was outside the time window to track outflow for limiting
+                    if (timeLimit <= block.timestamp - info.timestamp) {
+                        info.rate = 0;
+                    }
+                    //if the last transaction was within the time window, decreases the tracked outflow rate relative to the time elapsed, so that the limit is able to update in realtime rather than in blocks, making flows smooth, and increasing the rate available as time increases without a transaction
```

Comment:
this is a nice trick, but I think it's equivalent to spreading the recent outflow over the whole previous time window. Which in effect means that rateLimit as it's phrased can always be exceeded. E.g. if previous outflow was near the end of the window (e.g 59 minutes ago), the new outflow can almost be the size of `rateLimit` (because 59/60 of capacity are added) which means that the `rateLimit` strictly speaking is exceeded by a factor of almost two for the last hour.

I'm not sure if this means that actual rate limit is at most 2X the parameter for it (which is still limited), but probably needs to be mentioned in docs and description of the functionality.

Reply:
Have updated accordingly, will attach notice about implementation for how to set the and handle the limit being an advised plan not a requirement.